File: 572-subtree-of-another-tree/572-subtree-of-another-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def isSubtree(self, root, subRoot):
        
        
        
        if not subRoot:
            return True
        if not root:
            return False

        if self.isSame(root, subRoot):
            return True
        else:
            left = self.isSubtree(root.left, subRoot)
            right = self.isSubtree(root.right, subRoot)
            return left or right
        
        
        """if not subRoot and root:
            return False"""
        
        
        
        
    def isSame(self, root, subRoot):
        if not root and not subRoot:
            return True
            
        if root and subRoot and root.val == subRoot.val:
            left = self.isSame(root.left, subRoot.left)
            right = self.isSame(root.right, subRoot.right)
            return left and right
        else:
            return False

File: 572-subtree-of-another-tree/572-subtree-of-another-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {TreeNode} subRoot
 * @return {boolean}
 */
var isSubtree = function(root, subRoot) {
    if  (subRoot === null) return true
    if (root === null)return false

    if (isSame(root, subRoot)) return true
    if (!isSame(root, subRoot)){
        let left = isSubtree(root.left, subRoot)
        let right = isSubtree(root.right, subRoot)
        return left || right
    }
            
};
  
var isSame = (root, subRoot)=>{
    if (root === null && subRoot === null) return true
    if (root && subRoot && root.val == subRoot.val){
        let left = isSame(root.left, subRoot.left)
        let right = isSame(root.right, subRoot.right)
        return left && right
    }else{ return false }
}
        
            
        